src/components/hooks/useTodos.ts

```diff
@@ -0,0 +1,136 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import { Todo } from '../../types/Todo';
+import { ErrorType } from '../../types/ErrorType';
+import { fetchTodos } from '../../utils/fetchTodos';
+import { FilterType } from '../../types/FilterType';
+import { getFilteredTodos } from '../../utils/getFilteredTodo';
+import { todosService } from '../../api/todos';
+
+export const useTodos = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState<ErrorType>(
+    ErrorType.DEFAULT,
+  );
+  const [loadingTodosIds, setLoadingTodosIds] = useState<number[]>([]);
+  const [filter, setFilter] = useState(FilterType.all);
+
+  const timeoutRef = useRef<NodeJS.Timeout | null>(null);
+
+  const activeTodos = useMemo(
+    () => todos.filter(todo => !todo.completed),
+    [todos],
+  );
+  const completedTodos = useMemo(
+    () => todos.filter(todo => todo.completed),
+    [todos],
+  );
+
+  const handleRemoveError = useCallback(() => {
+    setErrorMessage(ErrorType.DEFAULT);
+  }, []);
+
+  const handleErrorMessage = useCallback(
+    (error: ErrorType) => {
+      setErrorMessage(error);
+
+      if (timeoutRef.current) {
+        clearTimeout(timeoutRef.current);
+      }
+
+      timeoutRef.current = setTimeout(() => {
+        handleRemoveError();
+        timeoutRef.current = null;
+      }, 3000);
+    },
+    [handleRemoveError],
+  );
+
+  const handleFilterChange = useCallback(
+    (filterBy: FilterType) => {
+      setFilter(filterBy);
+    },
+    [setFilter],
+  );
+
+  const filteredTodos = useMemo(
+    () => getFilteredTodos(todos, filter),
+    [todos, filter],
+  );
+
+  const deleteTodo = (todoId: number) => {
+    setLoadingTodosIds(current => [...current, todoId]);
+
+    return todosService
+      .delete(todoId)
+      .then(() => {
+        setTodos(current => current.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        handleErrorMessage(ErrorType.DELETE);
+      })
+      .finally(() => {
+        setLoadingTodosIds(current => current.filter(id => id !== todoId));
+      });
+  };
+
+  useEffect(() => {
+    fetchTodos(setTodos, handleErrorMessage);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```

Comment:
Move the `useEffect `to the end of code
Placing `useEffect `at the end improves readability by letting developers first grasp the main logic before seeing side effects. This promotes clearer structure and consistency in the codebase.